appservice/src/SiteWrapper.ts

```diff
@@ -116,6 +120,67 @@ export class SiteWrapper {
     }
 
+    public async localGitDeploy(fsPath: string, client: WebSiteManagementClient, outputChannel: vscode.OutputChannel, servicePlan: string): Promise<void> {
+        vscode.window.showInformationMessage('It updated to 2.2');
+        let taskResults: [User, SiteConfigResource];
+        const kuduClient: KuduClient = await this.getKuduClient(client);
+        const yes: string = 'Yes';
+        const pushReject: string = 'Push rejected due to Git history diverging. Force push?';
+
+        if (!this.slotName) {
+            // API calls for Web App
+            taskResults = await Promise.all([
+                client.webApps.listPublishingCredentials(this.resourceGroup, this.appName),
+                client.webApps.getConfiguration(this.resourceGroup, this.appName)
+            ]);
+        } else {
+            // API calls for Deployment Slot
+            taskResults = await Promise.all([
+                client.webApps.listPublishingCredentialsSlot(this.resourceGroup, this.appName, this.slotName),
+                client.webApps.getConfigurationSlot(this.resourceGroup, this.appName, this.slotName)
+            ]);
+        }
+
+        const publishCredentials: User = taskResults[0];
+        const config: SiteConfigResource = taskResults[1];
+        if (config.scmType !== 'LocalGit') {
+            // SCM must be set to LocalGit prior to deployment
+            await this.updateScmType(client, config);
+        }
+        // credentials for accessing Azure Remote Repo
+        const username: string = publishCredentials.publishingUserName;
+        const password: string = publishCredentials.publishingPassword;
+        const remote: string = `https://${username}:${password}@${this._gitUrl}`;
+        const localGit: git.SimpleGit = git(fsPath);
+        try {
+
+            const status: git.StatusResult = await localGit.status();
+            if (status.files.length > 0) {
+                const uncommit: string = `${status.files.length} uncommitted change(s) in local repo "${fsPath}"`;
+                vscode.window.showWarningMessage(uncommit);
+            }
+            await localGit.push(remote, 'HEAD:master');
+        } catch (err) {
+            // tslint:disable-next-line:no-unsafe-any
+            if (err.message.indexOf('spawn git ENOENT') >= 0) {
+                throw new errors.GitNotInstalledError();
+            } else if (err.message.indexOf('error: failed to push') >= 0) { // tslint:disable-line:no-unsafe-any
+                const input: string | undefined = await vscode.window.showErrorMessage(pushReject, yes);
+                if (input === 'Yes') {
+                    await (<(a: string, b: string, c: object) => Promise<void>>localGit.push)(remote, 'HEAD:master', { '-f': true });
+                    // Ugly casting neccessary due to bug in simple-git. Issue filed
+                } else {
+                    throw new errors.UserCancelledError();
```

Comment:
It's fine to use this error internally, but we don't want to 'expose' it as an error outside of the npm package. (After all - it's not _technically_ an error).

The typical pattern is to return 'undefined' if the user canceled (In other cases, you could return the deployment id or some other useful piece of information)